src/rules/no_const_assign.rs

```diff
@@ -0,0 +1,178 @@
+    }
+  }
+
+  fn check_array_pat(&mut self, array: &swc_ecma_ast::ArrayPat, span: Span) {
+    if !array.elems.is_empty() {
+      for elem in array.elems.iter() {
+        if let Some(element) = elem {
+          self.check_pat(element, span);
+        }
+      }
+    }
+  }
+
+  fn check_scope_for_const(&mut self, span: Span, ident: &str) {
+    let scope = self.scope_manager.get_scope_for_span(span);
+    if let Some(binding) = self.scope_manager.get_binding(scope, ident) {
+      if binding.kind == BindingKind::Const {
+        self.context.add_diagnostic(
+          span,
+          "no-const-assign",
+          "Reassigning constant variable is not allowed",
+        );
+      }
+    }
+  }
+}
+
+impl Visit for NoConstAssignVisitor {
+  fn visit_assign_expr(&mut self, assign_expr: &AssignExpr, _node: &dyn Node) {
+    match &assign_expr.left {
+      PatOrExpr::Expr(pat_expr) => {
+        if let Expr::Ident(ident) = &**pat_expr {
+          self.check_scope_for_const(assign_expr.span, &ident.sym.to_string());
+        }
+      }
+      PatOrExpr::Pat(boxed_pat) => self.check_pat(boxed_pat, assign_expr.span),
+    };
+  }
+
+  fn visit_update_expr(&mut self, update_expr: &UpdateExpr, _node: &dyn Node) {
+    if let Expr::Ident(ident) = &*update_expr.arg {
+      self.check_scope_for_const(update_expr.span, &ident.sym.to_string());
+    }
+  }
+}
+
+#[cfg(test)]
+mod tests {
+  use super::*;
+  use crate::test_util::*;
+
+  #[test]
+  fn no_const_assign_valid() {
+    assert_lint_ok::<NoConstAssign>(
+      r#"
+      const x = 0; { let x; x = 1; }
+      const x = 0; function a(x) { x = 1; }
+      const x = 0; foo(x);
+      for (const x in [1,2,3]) { foo(x); }
+      for (const x of [1,2,3]) { foo(x); }
+      const x = {key: 0}; x.key = 1;
+
+      // ignores non constant.
+      var x = 0; x = 1;
+      let x = 0; x = 1;
+      function x() {} x = 1;
+      function foo(x) { x = 1; }
+      class X {} X = 1;
+      try {} catch (x) { x = 1; }"#,
+    );
+  }
+
+  #[test]
+  fn no_const_assign_invalid() {
+    assert_lint_err::<NoConstAssign>("const x = 0; x = 1;", 13);
+    assert_lint_err::<NoConstAssign>("const {a: x} = {a: 0}; x = 1;", 23);
+    assert_lint_err::<NoConstAssign>("const x = 0; ({x} = {x: 1});", 15);
+    assert_lint_err::<NoConstAssign>("const x = 0; ({a: x = 1} = {});", 14);
+    assert_lint_err::<NoConstAssign>("const x = 0; x += 1;", 13);
+    assert_lint_err::<NoConstAssign>("const x = 0; ++x;", 13);
+    assert_lint_err::<NoConstAssign>(
+      "const x = 0; function foo() { x = x + 1; }",
+      30,
+    );
+    assert_lint_err::<NoConstAssign>(
+      "const x = 0; function foo(a) { x = a; }",
+      31,
+    );
+    assert_lint_err::<NoConstAssign>("for (const i = 0; i < 10; ++i) {}", 26);
+    assert_lint_err::<NoConstAssign>(
+      "const x = 0; while (true) { x = x + 1; }",
+      28,
+    );
+    assert_lint_err_n::<NoConstAssign>(
+      "const x = 0  ; x = 1; x = 2;",
+      vec![15, 22],
+    );
+  }
```

Comment:
Great tests! However it seems that scope analysis is still not fully done so I'd suggest to add test cases in all types of scopes. If some of them don't catch this violation - it's fine let's leave them commented out and start uncommenting as we progress on scope analysis. WDYT @disizali?

Reply:
I added new tests, they all worked as expected.
Since this rule respects all types of scopes and doesn't need to know the exact type of scope, the current state of the scope is sufficient for this one IMHO.

Reply:
Great to hear that!